### PR TITLE
Allow GET endpoint to use URL params instead of body

### PR DESCRIPTION
**Pull request by Blatoy**

### CODE REVIEW

File: dashboard/backend/api.js [hunk 1]
```diff
@@ -68,9 +68,28 @@ function createEndpointHandler(endpoint, { mustReturn = false } = {}) {
...
         Logger.log("[WebAPI] Accessing " + req.url);
 
-        // Strip trustedData off the userData
-        const appToken = req.body.appToken;
-        const userData = req.body.data;
+        let appToken;
+        const authorizationHeader = req.header("authorization");
+
+        if (typeof authorizationHeader === "string" && authorizationHeader.startsWith("Bearer ")) {
+            appToken = authorizationHeader.slice(7);
```
Comment (RedMser) [inline dashboard/backend/api.js]: Write a comment describing where the magic number 7 comes from.

File: dashboard/backend/api.js [hunk 2]
```diff
@@ -68,9 +68,28 @@ function createEndpointHandler(endpoint, { mustReturn = false } = {}) {
...
+        let userData = {};
+
+        if (req.method === "GET") {
+            // We except GET query params to be JSON encoded
+            try {
+                for (const key in req.query) {
+                    userData[key] = JSON.parse(req.query[key]);
+                }
+            } catch (err) {
+                reply({ status: 400, error: new WebApiError("GET endpoint excepts values to be JSON encoded", "get-params-not-json-encoded"), source: "core" });
```
Comment (RedMser) [inline dashboard/backend/api.js]: If easily possible, include the value of both `key` and `req.query[key]` in this error, so it's easy to know what caused the error.

File: dashboard/backend/api.js [hunk 3]
```diff
@@ -171,7 +190,9 @@ module.exports.registerActions = (app) => {
     app.get("/api/dashboard/test", createEndpointHandler({
         authentication: "loggedIn",
         action: (userData, trustedData) => {
-            console.log("lel", userData, trustedData);
+            console.log("IN ENDPOINT");
```
Comment (RedMser) [inline dashboard/backend/api.js]: :eyes:

File: dashboard/backend/api.js [hunk 4]
```diff
@@ -68,9 +68,28 @@ function createEndpointHandler(endpoint, { mustReturn = false } = {}) {
 
         Logger.log("[WebAPI] Accessing " + req.url);
 
-        // Strip trustedData off the userData
-        const appToken = req.body.appToken;
-        const userData = req.body.data;
+        let appToken;
+        const authorizationHeader = req.header("authorization");
```
Comment (RedMser) [inline dashboard/backend/api.js]: Uppercase **A** to be consistent with other HTTP headers and client code.

File: dashboard/www/scripts/utils/utils.js [hunk 5]
```diff
@@ -4,26 +4,51 @@ window.discotron.utils = class {
...
+            body = undefined; // Unset the body as get cannot have one
+        }
+
+        if (body !== undefined) {
+            request.body = JSON.stringify(body);
+        }
+
+        const response = await fetch(url, request);
+        try {
+            return await response.json();
```
Comment (RedMser) [inline dashboard/www/scripts/utils/utils.js]: Shouldn't this try block extend up to the `fetch` as well?

File: dashboard/www/scripts/utils/utils.js [hunk 6]
```diff
@@ -4,26 +4,51 @@ window.discotron.utils = class {
...
+        } catch (error) {
+            console.error("API Query failed for URL", response);
+            throw new discotron.WebApiError("An error occurred when communicating with the bot. Status code: " + response.status);
```
Comment (RedMser) [inline dashboard/www/scripts/utils/utils.js]: I'm kind of against using a WebApiError in a general `utils` function. Even though it's ugly and annoying, we should possibly re-catch on the calling side, and use a regular `Error` type here?

File: dashboard/www/scripts/utils/utils.js [hunk 7]
```diff
@@ -4,26 +4,51 @@ window.discotron.utils = class {
      * @param {string} verb HTTP verb to send request with
      * @param {string} url Url to make the post request on
      * @param {object} [body] Data that will be JSON.stringified and sent to the website. Does not work for GET requests.
```
Comment (RedMser) [inline dashboard/www/scripts/utils/utils.js]: Update this to reflect what happens for GET.

**Review by RedMser — CHANGES_REQUESTED** [review]

File: dashboard/backend/api.js (hunk 3 above)
Comment (Blatoy) [reply dashboard/backend/api.js]: :eyes:

File: dashboard/www/scripts/web-api.js [hunk 8]
```diff
@@ -1,3 +1,6 @@
+window.onerror = (a, b, c, d, e) => {
```
Comment (RedMser) [inline dashboard/www/scripts/web-api.js]: :eyes: 💦

**Review by RedMser — CHANGES_REQUESTED** [review]

**Pull request merged**